src/rhash/hash.c

```diff
@@ -2388,23 +2380,202 @@ static int rc_hash_gamecube(char hash[33], const rc_hash_iterator_t* iterator)
+    tmd_offset =
+      ((uint64_t)((quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3])) << 2;
+
+    if (tmd_size > CLUSTER_SIZE)
+    {
+      tmd_size = CLUSTER_SIZE;
+    }
+    rc_file_seek(iterator, file_handle, ((uint64_t)partition_table[jx] << 2) + tmd_offset, SEEK_SET);
+    rc_file_read(iterator, file_handle, buffer, tmd_size);
+    rc_hash_iterator_verbose_formatted(iterator, "Hashing %u byte title metadata (partition type %u)",
+                                       tmd_size, partition_table[jx + 1]);
+    md5_append(md5, buffer, tmd_size);
+
+    /* Hash partition */
+    rc_file_seek(iterator, file_handle, ((uint64_t)partition_table[jx] << 2) + 0x2B8, SEEK_SET);
+    rc_file_read(iterator, file_handle, quad_buffer, 4);
+    part_offset =
+      ((uint64_t)((quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3])) << 2;
+    rc_file_read(iterator, file_handle, quad_buffer, 4);
+    part_size =
+      ((uint64_t)((quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3])) << 2;
+
+    if (encrypted)
+    {
+      cluster_count = (part_size / 0x8000 > MAX_CLUSTER_COUNT) ? MAX_CLUSTER_COUNT : part_size / 0x8000;
+      rc_hash_iterator_verbose_formatted(iterator, "Hashing %u encrypted clusters (%u bytes)",
+                                         cluster_count, cluster_count * CLUSTER_SIZE);
+      buffer = (uint8_t*)malloc(CLUSTER_SIZE);
+      for (ix = 0; ix < cluster_count; ix++)
+      {
+        rc_file_seek(iterator, file_handle, part_offset + (ix * 0x8000) + 0x400, SEEK_SET);
+        rc_file_read(iterator, file_handle, buffer, CLUSTER_SIZE);
+        md5_append(md5, buffer, CLUSTER_SIZE);
+      }
+    }
+    else /* Decrypted */
+    {
+      if (rc_hash_nintendo_disc_partition(md5, iterator, file_handle, part_offset, 2) == 0)
+        return 0;
+    }
+  }
+  free(partition_table);
+  free(buffer);
+  return 1;
+}
+
+static int rc_hash_nintendo_disc(char hash[33], const rc_hash_iterator_t* iterator)
+{
+  md5_state_t md5;
+  void* file_handle;
+
+  uint8_t quad_buffer[4];
+  uint8_t success;
+
+  file_handle = rc_file_open(iterator, iterator->path);
+  if (!file_handle)
+    return rc_hash_iterator_error(iterator, "Could not open file");
+
+  md5_init(&md5);
+  /* Check Gamecube */
+  rc_file_seek(iterator, file_handle, 0x1c, SEEK_SET);
+  rc_file_read(iterator, file_handle, quad_buffer, 4);
+  rc_hash_iterator_verbose_formatted(iterator, "Checking Gamecube magic word: %u %u %u %u",
+                                     quad_buffer[0], quad_buffer[1], quad_buffer[2], quad_buffer[3]);
+  if (quad_buffer[0] == 0xC2 && quad_buffer[1] == 0x33 && quad_buffer[2] == 0x9F && quad_buffer[3] == 0x3D)
+  {
+    success = rc_hash_nintendo_disc_partition(&md5, iterator, file_handle, 0, 0);
+
+    rc_file_close(iterator, file_handle);
+
+    return rc_hash_finalize(iterator, &md5, hash);
```

Comment:
this close and return can be handled by the code after the else, which would also ensure the `success` value is checked.